feat(chat): 채팅 메시지에 파일 정보 추가

j1suk1m opened the pull request:
## 🐛 관련 이슈
 <!-- Closes #{이슈-번호} 형태로 작성하세요 (ex: Closes #134) -->
Closes #134 

## 📌 목적
- 채팅 메시지에 파일(`이미지`, `PDF`)을 포함할 수 있게 변경합니다.

## ✅ 변경 사항
### 1️⃣ [메시지 타입 추가](https://github.com/prgrms-be-adv-devcourse/beadv1_1_hexagon_BE/commit/b0d3dd3a10106c0cf41735561d18fb8388c50b18)

기존에는 메시지 타입이 별도로 존재하지 않았기 때문에, `TEXT`, `FILE`, `MIXED` enum을 추가했습니다.

- `TEXT`: 텍스트 메시지

- `FILE`: 파일 메시지

- `MIXED`: 텍스트 + 파일 메시지

<br>

### 2️⃣ [메시지 타입별 유효성 검증](https://github.com/prgrms-be-adv-devcourse/beadv1_1_hexagon_BE/blob/b0d3dd3a10106c0cf41735561d18fb8388c50b18/communication-service/src/main/java/com/example/communicationservice/type/MessageType.java)

현재 채팅 메시지 엔티티는 아래와 같이 구성되어 있습니다. 

```java
public class ChatMessage {

    @Id
    private String id; // MongoDB의 PK, ObjectId와 매핑

    private String roomId;

    private String senderCode;

    private MessageType type;

    private String text;

    private File file;

    @CreatedDate
    private Instant sentAt; // 전송된 시간

    ...
}
```
<br>

이때 메시지 타입에 따라 필드의 null 가능 여부가 달라집니다.

- `TEXT`: `text`는 `not null`, `file`은 `null`

- `FILE`: `text`는 `null`, `file`은 `not null`

- `MIXED`: `text`도 `not null`, `file`도 `not null`

<br>

이러한 유효성 검증은 단일 `DTO validation`만으로는 한계가 있기 때문에 `추상 메서드`를 활용했습니다.

1. 메시지 타입 enum 내부에 추상 메서드 `validate`를 정의합니다.

2. 각 enum 상수(`TEXT`, `FILE`, `MIXED`)가 해당 타입에 맞는 검증 로직을 오버라이드합니다.

3. 채팅 메시지 엔티티 생성 시 생성자 내에서 `validate`를 호출해 유효성을 검증합니다.

<br>

이 방식은 다음과 같은 장점이 있습니다.

- 타입별 검증 로직이 enum 내부에 응집되어 있어 `가독성`이 향상됩니다.

- 서비스 레이어에서 검증 분기를 수행할 필요가 없기 때문에 `비즈니스 로직이 단순`해집니다.

- 새로운 메시지 타입이 추가될 경우 새 enum 상수를 정의하고 검증 로직을 오버라이드하면 되므로 `확장성`이 증가합니다.

<br>

### 🛠️ [메시지 타입별 유효성 검증 리팩토링](https://github.com/prgrms-be-adv-devcourse/beadv1_1_hexagon_BE/pull/174/commits/f10b81bb5cd0b1cb2557ecba5ae5d18e53c40e4f)

초기 구현은 각 enum 상수(`TEXT`, `FILE`, `MIXED`)가  `validate`를 오버라이드하는 방식이었습니다.

그러나 중복되는 검증 코드가 각 타입에 흩어져 있는 문제가 있었기 때문에,

인수님(@ZZAMBAs)의 피드백을 반영하여 함수형 기반의 `Consumer` 구조로 리팩토링했습니다.

1. 각 메시지 타입은 자신이 필요로 하는 검증 조합을 `Consumer` 체인 형태로 정의합니다.

2. 공통적으로 사용되는 세부 검증 로직을 enum 아래 정적 메서드로 모듈화합니다.

3. `validate`는 단순히 해당 타입의 `validator`를 호출하는 역할만 하도록 축소했습니다.

<br>

리팩토링된 예시는 아래와 같습니다.

```java
TEXT(message -> {
    requireText(message);
    validateTextLength(message);
    forbidFile(message);
}),
FILE(message -> {
    requireFile(message);
    requireFileKey(message);
    forbidText(message);
}),
MIXED(message -> {
    requireText(message);
    validateTextLength(message);
    requireFile(message);
    requireFileKey(message);
});
```

<br>

이 방식으로 변경하면서 다음과 같은 효과가 있었습니다.

- 공통 검증 로직이 한 곳에 모여 정리되어 있어 `유지보수성`이 크게 향상되었습니다.

- 각 타입이 필요로 하는 검증 흐름을 순서대로 나열하는 구성 방식이 되어 `가독성`이 개선되었습니다.

- 새로운 메시지 타입이 추가될 경우 필요한 검증 메서드를 조합하면 되므로 `확장`과 `변경`이 훨씬 유연해졌습니다.

<br>

### 3️⃣ [DTO <-> 엔티티 변환 mapper 추가](https://github.com/prgrms-be-adv-devcourse/beadv1_1_hexagon_BE/commit/781eddb26f39f17ea7c5397fc72e761ae0141e73)

기존에는 DTO 내에 `정적 팩토리 메서드`를 구현해 DTO <-> 엔티티 변환을 수행했기 때문에, DTO가 엔티티를 의존하는 문제가 있었습니다.

```java
public record ChatRoomCreateResponse(
    String id
) {
    public static ChatRoomCreateResponse from(ChatRoom chatRoom) { // 엔티티 의존
        return new ChatRoomCreateResponse(chatRoom.getId());
    }
}
```

<br>

따라서 레이어 간 결합도를 낮추기 위해 별도의 mapper를 구현했습니다.

```java
// dto와 entity 사이의 변환 로직 전담
public abstract class ChatMapper {

    private ChatMapper() {} // 인스턴스화 방지

    // ------------------ ChatMessage ------------------

    public static ChatMessage toEntity(ChatMessageSendRequest request) {
        return ChatMessage.builder()
            .roomId(request.roomId())
            .senderCode(request.senderCode())
            .type(request.type())
            .text(request.text())
            .file(request.file() != null ? toEntity(request.file()) : null)
            .build();
    }

    ...
}
```

## 🧪 테스트 방법
기존 서비스 단위 테스트 코드에 변경 사항을 반영했습니다.

## 📎 참고 사항
- 관련 이슈 번호 : #134 

## 📝 제출자 검토 리스트

- [x] 실행 시 오류가 없나요?
- [x] 테스트 전부 통과 했나요?
- [x] 이슈 번호를 입력 했나요?
- [x] 코드 컨벤션을 모두 지켰나요?
- [x] 브랜치가 develop에 merge 요청을 보냈나요?

## 📝 검토자 체크 리스트

> [!IMPORTANT]
> 직접 모든 부분에 대해 검토하고 멘션이나 이모티콘 남겨주세요

- [ ] 요청을 받은 후 하루 이내에 피드백을 보내주세요
- [ ] 본인이 검토해야하는 PR 피드백을 주세요.
- [ ] 브랜치가 develop에 merge 요청을 보냈나요?

j1suk1m committed:
feat(chat): 채팅 메시지 타입 enum 추가

- 기존: 텍스트 메시지만 가능
- 현재: 텍스트(TEXT)/파일(FILE)/혼합(MIXED) 메시지 가능
- enum 내부에 메시지 타입별 검증 로직 구현
- 검증 실패 코드 추가
- 커스텀 예외 추가

j1suk1m committed:
feat(chat): 파일 관련 class 추가

- ChatMessage에 포함되는 embedded document class 추가
- 요청 및 응답에 사용되는 dto 추가

j1suk1m committed:
refactor(chat): entity에 변경 사항 반영

- 채팅 메시지 타입 enum 추가
- 파일 embedded document 추가
- 메시지 타입을 세분화하기 위해 기존의 content 필드를 text로 변경

j1suk1m committed:
refactor(chat): dto에 변경 사항 반영

- 채팅 메시지 타입 enum 추가
- 파일 embedded document의 dto 추가
- content 필드를 text로 변경

j1suk1m committed:
refactor(chat): dto <-> entity 변환 mapper 구현

- 기존: dto 내부에서 entity 변환 수행
- 현재: mapper에서 dto <-> entity 변환 전담

j1suk1m committed:
refactor(chat): service/controller에 변경 사항 반영

- dto 대신 mapper의 메서드를 호출하도록 변경
- service가 request dto를 받고, response dto를 반환하도록 변경
- controller에 `@Valid` 적용 -> 요청 dto 유효성 검증 활성화

j1suk1m committed:
test(chat): test에 변경 사항 반영

ZZAMBAs commented on communication-service/src/main/java/com/example/communicationservice/mapper/ChatMapper.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    // ------------------ Paging ------------------
+
+    public static PageInfo toPageInfo(Page<?> page) {
+        return new PageInfo(
+            page.getNumber(),
+            page.getSize(),
+            page.getTotalElements(),
+            page.getTotalPages()
+        );
+    }
+
+    // ------------------ File ------------------
+
+    public static FileInfo from(File file) {
+        if (file == null) {
+            return null;
+        }
+
+        return new FileInfo(file.getKey());
+    }
+
+    public static File toEntity(FileInfo fileInfo) {
+        if (fileInfo == null) {
+            return null;
+        }
+
+        return File.builder()
+            .key(fileInfo.key())
+            .build();
+    }
```
이 변환은 범용적인 느낌이 들어서 따로 클래스 분리를 해도 괜찮아 보여요

programofktw replied:
저도 분리해도 좋을 것 같지만 
Communication module에서 Mapper는 이것뿐일 것 같기두해서 그대로 둬도 괜찮을 것 같습니다!

j1suk1m replied:
좋은 의견 감사합니다! 말씀해주신 것처럼 Page나 File 관련 매핑 로직은  
ChatMapper의 역할 범위를 벗어날 여지가 있다고 생각합니다🤔

다만 현재는 파일 관련 기능 규모가 크지 않아 우선은 ChatMapper 내에서 유지하게 되었습니다.

다음 PR에서 파일 메시지 전송 기능을 확장하면서 범용 mapper로 분리하는 방향도 함께 고려해보겠습니다!

ZZAMBAs commented on communication-service/src/main/java/com/example/communicationservice/type/MessageType.java:
열거형에 로직을 둔 좋은 코드 같습니다. 간결화를 위해 `Consumer<T>` 람다식을 이용해도 괜찮을 것 같습니다. 그리고 공통 검증을 따로 빼면 내부 코드도 간결해 질 것 같아요.

j1suk1m replied:
가독성, 응집도, 확장성 측면에서 더 좋겠다는 생각이 들어 반영해보겠습니다!

ZZAMBAs commented on communication-service/src/main/java/com/example/communicationservice/controller/dto/response/PageInfo.java:
`hasNext` 값을 함께 전달해 준다면 마지막 페이지일 때 프론트엔드에서 API 호출을 하지 않도록 할 수 있을 것 같아요. 
`Page<T>`에서도 `hasNext`가 있는 것으로 알고 있어요!

j1suk1m replied:
맞는 말씀입니다! 다른 코드 변경 없이 dto와 mapper만 수정하면 돼서 너무 좋네요👓

ZZAMBAs reviewed (approved):
고생하셨습니다!

programofktw reviewed (approved):
내용 상 문제되는 부분은 없는 것 같습니다!
코드를 간결하고 가독성 좋게 짜주셔서 빠르게 리뷰할 수 있었습니다.
저도 비슷하게 할 수 있게 노력해야겠네요.

추가적으로 저도 이번에 테스트 코드를 작성하면서 알게 된 내용으로
Test의 이름(method 명 혹은 @ DisplayName() 설정)을 지어줄 때

테스트 관점이 아닌 실제 동작 관점에서 작성해주는 것이 좋다는 이야기를 보았습니다.
저도 최근 작업에서만 반영을 하고 있는데
지수님 PR에서 예시를 들어보자면

채팅방_참여자_수가_2명이_아니면_채팅방_생성에_실패한다

의 경우

채팅방 생성 요청 시 채팅방 참여자 수가 2명이 아니라면 CHATROOM_INVALID_MEMBER_COUNT ErrorCode가 발생한다

와 같이 써주는 방식 써주면 이후 다른 개발자가 테스트 이름만 보고도 어떤 테스트인지 확인하기 쉽다 같은 느낌의 정보였습니다.

실제 동작과 무관한 부분이기에
단순 정보 공유 관점에서 Comment를 작성해보았습니다.

j1suk1m committed:
refactor(chat): 채팅 메시지 타입 enum 유효성 검증 구조 개선

- 검증 로직을 Consumer 기반 함수형 구조로 변경
- 공통 검증 로직 분리로 중복 제거 및 응집도 강화
- 검증 로직을 규칙 중심 구조로 개선해 가독성 향상

j1suk1m committed:
feat(chat): 페이지 응답 dto에 hasNext 필드 추가

- Page<T>의 hasNext 값을 페이지 응답 dto에 포함
- 프론트엔드가 마지막 페이지 여부를 쉽게 판단하도록 개선
